docker/archive/dest.go

```diff
@@ -19,15 +19,24 @@ func newImageDestination(ctx *types.SystemContext, ref archiveReference) (types.
 	if ref.destinationRef == nil {
 		return nil, errors.Errorf("docker-archive: destination reference not supplied (must be of form <path>:<reference:tag>)")
 	}
-	fh, err := os.OpenFile(ref.path, os.O_WRONLY|os.O_EXCL|os.O_CREATE, 0644)
+
+	// ref.path can be either a pipe or a regular file
+	// in the case of a pipe, we require that we can open it for write
+	// in the case of a regular file, we don't want to overwrite any pre-existing file
+	// so we check for Size() == 0 below (This is racy, but using O_EXCL would also be racy,
+	// only in a different way. Either way, it’s up to the user to not have two writers to the same path.)
+	fh, err := os.OpenFile(ref.path, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening file %q", ref.path)
+	}
+
+	fhStat, err := fh.Stat()
 	if err != nil {
-		// FIXME: It should be possible to modify archives, but the only really
-		//        sane way of doing it is to create a copy of the image, modify
-		//        it and then do a rename(2).
-		if os.IsExist(err) {
-			err = errors.New("docker-archive doesn't support modifying existing images")
-		}
-		return nil, err
+		return nil, errors.Wrapf(err, "error statting file %q", ref.path)
+	}
+
+	if fhStat.Mode()&os.ModeType != os.ModeNamedPipe && (fhStat.Mode().IsRegular() && fhStat.Size() != 0) {
```

Comment:
This is not actually a mathematical negation of the original, and `fhStat.Mode().IsRegular()` always implies `fhStat.Mode()&os.ModeType != os.ModeNamedPipe`.

We could either just ignore other file types, and say
```go
if fhStat.Mode().IsRegular() && fhStat.Size() != 0 {
   /* fail */
}
```
, or something like
```go
if fhStat.Mode().IsRegular() && fhStat.Size() != 0 {
  /* fail as above */
} else if fhStat.Mode()&os.ModeType != os.ModeNamedPipe {
  return nil, fmt.Errorf("Unsupported file type with mode %x", fhstat.Mode()")
}
```

Looking at the history of this PR, all we really wanted to check for is the non-zero size, so the shorter variant should be enough.

I guess I can live with the code as is, so :+1: in the interest of speeding things up. But I’d prefer this to be cleaned up—and we are waiting for @runcom anyway.